packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,152 @@
+import { ISemver } from "src/universal/ISemver.sol";
+import { IL2ToL2CrossDomainMessenger } from "src/L2/IL2ToL2CrossDomainMessenger.sol";
+
+/// @custom:proxied
+/// @custom:predeploy 0x4200000000000000000000000000000000000023
+/// @title L2ToL2CrossDomainMessenger
+/// @notice The L2ToL2CrossDomainMessenger is a higher level abstraction on top of the CrossL2Inbox that provides
+///         features necessary for secure transfers ERC20 tokens between L2 chains. Messages sent through the
+///         L2ToL2CrossDomainMessenger on the source chain receive both replay protection as well as domain binding.
+contract L2ToL2CrossDomainMessenger is IL2ToL2CrossDomainMessenger, ISemver {
+    /// @notice Storage slot for the sender of the current cross domain message.
+    ///         Equal to bytes32(uint256(keccak256("l2tol2crossdomainmessenger.sender")) - 1)
+    bytes32 public constant CROSS_DOMAIN_MESSAGE_SENDER_SLOT =
+        0xb83444d07072b122e2e72a669ce32857d892345c19856f4e7142d06a167ab3f3;
+
+    /// @notice Storage slot for the source of the current cross domain message.
+    ///         Equal to bytes32(uint256(keccak256("l2tol2crossdomainmessenger.source")) - 1)
+    bytes32 public constant CROSS_DOMAIN_MESSAGE_SOURCE_SLOT =
+        0x711dfa3259c842fffc17d6e1f1e0fc5927756133a2345ca56b4cb8178589fee7;
+
+    /// @notice Current message version identifier.
+    uint16 public constant MESSAGE_VERSION = uint16(0);
+
+    /// @custom:semver 1.0.0
+    string public constant version = "1.0.0";
+
+    /// @notice Mapping of message hashes to boolean receipt values. Note that a message will only
+    ///         be present in this mapping if it has successfully been relayed on this chain, and
+    ///         can therefore not be relayed again.
+    mapping(bytes32 => bool) public successfulMessages;
+
+    /// @notice Nonce for the next message to be sent, without the message version applied. Use the
+    ///         messageNonce getter which will insert the message version into the nonce to give you
+    ///         the actual nonce to be used for the message.
+    uint240 internal msgNonce;
+
+    /// @notice Emitted whenever a message is sent to the other chain.
+    /// @param data Encoded data of the message that was sent.
+    event SentMessage(bytes data) anonymous;
+
+    /// @notice Emitted whenever a message is successfully relayed on this chain.
+    /// @param msgHash Hash of the message that was relayed.
+    event RelayedMessage(bytes32 indexed msgHash);
+
+    /// @notice Emitted whenever a message fails to be relayed on this chain.
+    /// @param msgHash Hash of the message that failed to be relayed.
+    event FailedRelayedMessage(bytes32 indexed msgHash);
+
+    /// @notice Retrieves the sender of the current cross domain message.
+    /// @return _sender Address of the sender of the current cross domain message.
+    function crossDomainMessageSender() external view returns (address _sender) {
```

Comment:
We should do the same thing here where these functions revert if the called flag in tstorage isn't set to differentiate between 0 and nil